scripts/install-nix-from-closure.sh

```diff
@@ -1,149 +1,618 @@
 #!/bin/sh
+### v 1.11.15-1
+###
+###############################
+###
+### Nix installation script
+### Universal shell script for for OS Linux environments
+###
+###############################
+###
+### Upstream URL: https://github.com/NixOS/nix/blob/master/scripts/install-nix-from-closure.sh
+###
+### Script is fully POSIX compatible, with controlled exceptions.
+### (must be) Fully transnational. That means, if it not succeeded - it does rolls back.
+###
+### `/bin/sh -u` is not possible to do, because Docker environments has USER='', and shell checks that and not get even to execution part
+###
+#
+###############################
+###  Documentation
+###############################
+{
+#
+#      Installer consist of:
+#      1. Setup environment
+#      2. Main constants
+#      3. CLI control constants
+#      4. CLI output functions
+#      5. Program stage functions
+#      6. Main function
+#
+#
+
+# Special things about this script
+#
+# 1) Notice, Warning, Error, ErrorRevert massages have special form to be very readable and get attention.
+# Message body starts from new line.
+# And has 4 spaces from the left. Always.
+#
+###############################
+#
+#    ^ notice "
+#    |
+#<---Install executed for ROOT.
+#
+#    Doing classic Linux package manager mode.
+#    In Nix this mode is called: single-user mode for root.
+#
+#    Nix can do multi-user mode, and manage package trees for users independently.
+#    "
+#    fi
+#
+###############################
+#
+# This produces output in color:
+#
+###############################
+#
+#./install: Notice:
+#
+#    Install executed for ROOT.
+#
+#    Doing classic Linux package manager mode.
+#    In Nix this mode is called: single-user mode for root.
+#
+#    Nix can do multi-user mode, and manage package trees for users independently.
+#
+###############################
+#
+true # so {...} is not empty, shell will not permit otherwise
+}
+###############################
+###  Setup environment
+###############################
+{
+# Set the character collating sequence to be numeric ASCII/C standard.
+# Set the character set to be standard one-byte ASCII.
+# Set default umask; to be non-restrictive and friendly to others.
+#
+LC_COLLATE=C ; export LC_COLLATE
+LANG=C ; export LANG
+umask 022
+}
+###############################
+###  Main constants
+###############################
+{
+readonly dest="/nix"
+readonly self="$(dirname "$0")"
+readonly nix="@nix@"
+readonly cacert="@cacert@"
+readonly appname="$0"
+}
+###############################
+###  CLI control constants
+###############################
+{
+# tput is not present on some systems (Alpine Linux), this trick with printf allows to store, not 'codes' - literal symbols
+readonly red=$(printf '\033[1;31m')
+readonly green=$(printf '\033[1;32m')
+readonly yellow=$(printf '\033[1;33m')
+readonly blue=$(printf '\033[1;34m')
+readonly bold=$(printf '\033[1m')
+readonly reset=$(printf '\033[0;m') # Reset to default output
```

Comment:
Yes you are right, Alpine for example not provide terminfo and `tput`.
So it can be that other systems not provide also.

Since Alpine is widely used now (in Docker containers), I used color codes.

`'\033[1;33m'` is more portable than '\e[1;33m' and HEX version.

And in this way I grab color codes is the most convenient way I found to use them.

Current output functions output color codes if terminal if color codes not supported.
But as does current master:
```shell
printf '\e[1;31mSwitching to the Multi-User Darwin Installer\e[0m\n'
...
printf '\e[1;31mwarning: installing Nix as root is not supported by this script!\e[0m\n'
```
And nobody objected to that.

Current approach is KISS and covers almost all setups.

I see further feature possible - is to check `terminfo` information, and availability of `tput` and usage of it - is posssible. And if `tput` is not present - color codes. And if terminal says it does not support color - does not use them at all. But we see now, - this is complication, that needs to be worked separately, if worked on.

Reply:
> Current output functions output color codes if terminal if color codes not supported.
> But as does current master:
> printf '\e[1;31mwarning: installing Nix as root is not supported by this script!\e[0m\n'
> And nobody objected to that.

It'd be nice to fix at some point though. I sometimes look at build logs from Jenkins where I'm installing and invoking Nix and I don't like what I see.

Reply:
I can work on it.
It is a topic that puzzled my mind for some time, how to do that great.
I even created and populated (this ArchWiki page)[https://wiki.archlinux.org/index.php/Color_output_in_console]. And studied the convoluted history and terminology of text output in Unix/Linux systems.

At start, I need to get the process going at least.

Reply:
I think as a start, one can assume that all terminals supports color and all non-terminals do not. At least that would prevent getting ansi escape codes in logfiles etc. Just ```test -t 1 && echo "use colors"```.

Reply:
Yeah.
I understand. With this version color codes going to show more.
There is Jenkins plugin that support [ANSI Colors](https://wiki.jenkins.io/display/JENKINS/AnsiColor+Plugin), maybe that helps.

I think it is possible to be something beyond lines:
```shell
if tput colors ; then
    # do color coding catch with tput
elif
    if [ -z $(printf '\033[1;32m') ] ; then
        # do color coding catch with ANSI codes
    fi
fi
```

Reply:
But as I said, lets not feature creap. And this idea is work after.
Since stable release was recently, this it dev master window. So spamming of color codes for a week is a good trade-off for smooth Linux installation process.

Reply:
@bjornfor 
So what do you say we choose:
To say to people: "plugins" - or `test -t 1` and let them have grim text 8) ?

Reply:
printf test is just my momentarily idea, brushstroke to explore. I mean to test, does terminal prints color codes, or not.

Yes, `if [ -t ]` is classic and old-school, if being serious, it is right thing to implement.

Reply:
Please do, it's only a couple of lines: test for `-t 1`, then `type -p tput` and then what you have now. I would think the printf test just gives you a string with the escape codes, since the terminal doesn't intercept anything.

Reply:
@wmertens: Better use "command -v" instead of "type -p" if you want to be compatible with POSIX sh.

Reply:
`which` is side binary.
`type -p` not in standard (http://pubs.opengroup.org/onlinepubs/9699919799/)
`command -v` is indeed in POSIX standard: http://pubs.opengroup.org/onlinepubs/9699919799/

Used `command -v` long ago myself, forgotten all things.
So, `which` -> `command -v` is done. In: https://github.com/NixOS/nix/pull/1565/commits/e2b9650833e45fba2a8b70086e49b71c6e60624b

Reply:
`tput` in standard has only `-T` and `init`, `clear` and `reset`.
Than we have case. `terminfo` and `termcap` `tput` implementations have different commands.
Real `termcap` is probably not used anywhere now, but maybe... I not found any list or mentioning of such systems. Beside latest canonical Berkeley Software Distribution 4.4, but it is 22 years old.

Reply:
Done. In: https://github.com/NixOS/nix/pull/1565/commits/953fbde5ea35d06b5f028c5c6e428d71286d9efb
```shell
if test -t ; then    # File descriptor is associated with a terminal - output colors
    if tput colors > /dev/null 2>&1 ; then
        # use tput and terminfo DB
        readonly red=$(tput setaf 1)
        readonly green=$(tput setaf 2)
        readonly yellow=$(tput setaf 3)
        readonly blue=$(tput setaf 4)
        readonly bold=$(tput smso)
        readonly reset=$(tput sgr0) # Reset to default output
    else
        # tput is not present on some systems (Alpine Linux), this trick with printf allows to store, not 'codes' - literal symbols
        readonly red=$(printf '\033[1;31m')
        readonly green=$(printf '\033[1;32m')
        readonly yellow=$(printf '\033[1;33m')
        readonly blue=$(printf '\033[1;34m')
        readonly bold=$(printf '\033[1m')
        readonly reset=$(printf '\033[0;m') # Reset to default output
    fi
fi
```

But, it needs to be looked, does in cases: `tput colors`: 256, 16, 8 - it outputs those colors. It should in all. And color codes also should.

Also it is probably possible, if no `terminfo` `tput colors` to test for `termcap` with `tput Co`.

If someone knows something in that regard - I am happy to know.

This one still KISS & covers almost all installs.